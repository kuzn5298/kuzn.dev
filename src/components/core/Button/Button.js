/* eslint-disable react/require-default-props */
import React from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';
import { ButtonContainer } from './Button.styled';

const Button = ({
    startIcon: StartIcon = null,
    endIcon: EndIcon = null,
    children,
    size = 'medium',
    border = false,
    className,
    ...props
}) => (
    <ButtonContainer size={size} className={clsx(border && 'border', className)} {...props}>
        {StartIcon}
        <span>{children}</span>
        {EndIcon}
    </ButtonContainer>
);

Button.propTypes = {
    startIcon: PropTypes.element,
    endIcon: PropTypes.element,
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    border: PropTypes.bool,
};

export default Button;
