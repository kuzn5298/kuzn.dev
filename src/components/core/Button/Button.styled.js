import styled, { css } from 'styled-components';

const sizeBlock = css`
    ${({ size }) => {
        switch (size) {
            case 'small':
                return css`
                    gap: 0.5rem;
                    padding: 0.25rem 0.75rem;
                    font-size: 0.75rem;
                `;
            case 'large':
                return css`
                    gap: 1.5rem;
                    padding: 0.75rem 2.25rem;
                    font-size: 1rem;
                `;
            default:
                return css`
                    gap: 1rem;
                    padding: 0.5rem 1.5rem;
                    font-size: 1rem;
                `;
        }
    }}
`;

export const ButtonContainer = styled.button`
    ${sizeBlock}

    position: relative;
    border: none;
    cursor: pointer;
    box-sizing: border-box;
    display: inline-flex;
    vertical-align: middle;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    background-color: transparent;
    color: ${({ theme }) => theme.palette.text.primary};
    text-transform: uppercase;
    line-height: 1;

    &::after,
    &:before {
        box-sizing: border-box;
        content: '';
        display: block;
        position: absolute;
        width: 30%;
        height: 50%;
        border: 0.125em solid;
        transition: all 0.6s ease;
    }

    &::after {
        bottom: 0;
        right: 0;
        border-top-color: transparent;
        border-left-color: transparent;
        border-bottom-color: currentColor;
        border-right-color: currentColor;
    }

    &::before {
        top: 0;
        left: 0;
        border-bottom-color: transparent;
        border-right-color: transparent;
        border-top-color: currentColor;
        border-left-color: currentColor;
    }

    &.border:after,
    &.border:before,
    &:hover:after,
    &:hover:before {
        width: 100%;
        height: 100%;
    }
`;
