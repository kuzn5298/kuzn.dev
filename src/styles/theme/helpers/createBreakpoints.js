const STEP = 5 / 100;

const createBreakpoints = (values, unit = 'px') => {
    const keys = Object.keys(values);

    const up = (key) => {
        const value = values[key] ?? key;
        return `@media (min-width: ${value}${unit})`;
    };

    const down = (key) => {
        const value = values[key] ?? key;
        return `@media (max-width: ${value - STEP}${unit})`;
    };

    return {
        keys,
        values,
        up,
        down,
        unit,
    };
};

export default createBreakpoints;
