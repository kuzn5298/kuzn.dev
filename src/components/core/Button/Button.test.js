import React from 'react';
import { wrapWithTheme } from '@tests/helpers';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/react';
import Button from './Button';

describe('<Button />', () => {
    it('should render component', () => {
        const { getByRole } = render(wrapWithTheme(<Button>Click me</Button>));
        expect(getByRole('button')).toBeInTheDocument();
    });

    it('should render component with default props', () => {
        const { getByRole } = render(wrapWithTheme(<Button>Click me</Button>));
        const buttonElement = getByRole('button');

        expect(buttonElement).not.toHaveClass('border');
        expect(buttonElement).toHaveTextContent('Click me');
        expect(buttonElement).toHaveStyle({
            gap: '1rem',
            padding: '0.5rem 1.5rem',
            fontSize: '1rem',
        });
    });

    it('should renders start and end icons', () => {
        const StartIcon = <svg data-testid="start-icon" />;
        const EndIcon = <svg data-testid="end-icon" />;

        const { getByTestId } = render(
            wrapWithTheme(
                <Button startIcon={StartIcon} endIcon={EndIcon}>
                    Click me
                </Button>
            )
        );

        expect(getByTestId('start-icon')).toBeInTheDocument();
        expect(getByTestId('end-icon')).toBeInTheDocument();
    });

    it('should render a correct button size', () => {
        const { getByTestId } = render(
            wrapWithTheme(
                <>
                    <Button size="small" data-testid="small-button">
                        Click me
                    </Button>
                    <Button size="medium" data-testid="medium-button">
                        Click me
                    </Button>
                    <Button size="large" data-testid="large-button">
                        Click me
                    </Button>
                </>
            )
        );

        expect(getByTestId('small-button')).toHaveStyle({
            gap: '0.5rem',
            padding: '0.25rem 0.75rem',
            fontSize: '0.75rem',
        });
        expect(getByTestId('medium-button')).toHaveStyle({
            gap: '1rem',
            padding: '0.5rem 1.5rem',
            fontSize: '1rem',
        });
        expect(getByTestId('large-button')).toHaveStyle({
            gap: '1.5rem',
            padding: '0.75rem 2.25rem',
            fontSize: '1rem',
        });
    });

    it('should render a correct button with border', () => {
        const { getByTestId } = render(
            wrapWithTheme(
                <Button border data-testid="border-button">
                    Click me
                </Button>
            )
        );

        expect(getByTestId('border-button')).toHaveClass('border');
    });

    it('should render a button with different props (snapshot test)', () => {
        const StartIcon = <svg>start icon</svg>;
        const EndIcon = <svg>end icon</svg>;

        const { asFragment } = render(
            wrapWithTheme(
                <Button startIcon={StartIcon} endIcon={EndIcon} size="large" border>
                    Click me
                </Button>
            )
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onClick handler when clicked (event)', async () => {
        const handleClick = jest.fn();
        const { getByRole } = render(
            wrapWithTheme(<Button onClick={handleClick}>Click me</Button>)
        );

        await userEvent.click(getByRole('button'));

        expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('should focus the button on Tab press (event)', async () => {
        const handleClick = jest.fn();
        const { getByRole } = render(
            wrapWithTheme(<Button onClick={handleClick}>Click me</Button>)
        );

        const buttonElement = getByRole('button');

        expect(buttonElement).not.toHaveFocus();

        await userEvent.tab();
        expect(buttonElement).toHaveFocus();
    });

    it('should call onClick on Enter and Space key (event)', async () => {
        const handleClick = jest.fn();
        const { getByRole } = render(
            wrapWithTheme(<Button onClick={handleClick}>Click me</Button>)
        );

        const buttonElement = getByRole('button');
        buttonElement.focus();

        await userEvent.keyboard('{Enter}');
        expect(handleClick).toHaveBeenCalledTimes(1);

        await userEvent.keyboard('{ }');
        expect(handleClick).toHaveBeenCalledTimes(2);
    });
});
