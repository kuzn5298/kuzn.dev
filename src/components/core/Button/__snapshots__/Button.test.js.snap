// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a button with different props (snapshot test) 1`] = `
<DocumentFragment>
  .c0 {
  gap: 1.5rem;
  padding: 0.75rem 2.25rem;
  font-size: 1rem;
  position: relative;
  border: none;
  cursor: pointer;
  box-sizing: border-box;
  display: inline-flex;
  vertical-align: middle;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  background-color: transparent;
  color: #1D1C1C;
  text-transform: uppercase;
  line-height: 1;
}

.c0::after,
.c0:before {
  box-sizing: border-box;
  content: '';
  display: block;
  position: absolute;
  width: 30%;
  height: 50%;
  border: 0.125em solid;
  transition: all 0.6s ease;
}

.c0::after {
  bottom: 0;
  right: 0;
  border-top-color: transparent;
  border-left-color: transparent;
  border-bottom-color: currentColor;
  border-right-color: currentColor;
}

.c0::before {
  top: 0;
  left: 0;
  border-bottom-color: transparent;
  border-right-color: transparent;
  border-top-color: currentColor;
  border-left-color: currentColor;
}

.c0.border:after,
.c0.border:before,
.c0:hover:after,
.c0:hover:before {
  width: 100%;
  height: 100%;
}

<button
    class="c0 border"
  >
    <svg>
      start icon
    </svg>
    <span>
      Click me
    </span>
    <svg>
      end icon
    </svg>
  </button>
</DocumentFragment>
`;
